src/appfl/run_mpi_async.py

```diff
@@ -0,0 +1,354 @@
+    # dest is the rank of the receiver and tag = dest + comm_size 
+    # we use different tags here to differentiate different types of messages (INFO v.s. MODEL)
+    send_reqs = [comm.Isend(np.frombuffer(global_model_bytes, dtype=np.byte), dest=i, tag=i+comm_size) for i in range(1, num_clients+1)]
+
+    # Wait for response (buffer size) - INFO - from clients
+    recv_reqs = [comm.irecv(source=i, tag=i) for i in range(1, num_clients+1)]
+
+    # FedAsync: main global training loop
+    test_loss = 0.0
+    test_accuracy = 0.0
+    best_accuracy = 0.0
+    while True:
+        # Wait for response from any one client
+        client_idx, local_model_size = MPI.Request.waitany(recv_reqs)
+
+        if client_idx != MPI.UNDEFINED:
+            # Record time
+            local_start_time = client_local_time[client_idx]
+            local_update_time = time.time() - client_local_time[client_idx]
+            global_update_start = time.time()
+
+            # Increment the global step
+            global_step += 1
+            logger.info(f"[Server Log] [Step #{global_step:3}] Server gets model size from client #{client_idx}")
+            
+            # Allocate a buffer to receive the model byte stream
+            local_model_bytes = np.empty(local_model_size, dtype=np.byte)
+
+            # Receive the model byte stream
+            comm.Recv(local_model_bytes, source=client_idx+1, tag=client_idx+1+comm_size)
+            logger.info(f"[Server Log] [Step #{global_step:3}] Server gets model from client #{client_idx}")
+
+            # Load the model byte to state dict
+            local_model_buffer = io.BytesIO(local_model_bytes.tobytes())
+            local_model_dict = torch.load(local_model_buffer)
+            
+            # Perform global update
+            logger.info(f"[Server Log] [Step #{global_step:3}] Server updates global model based on the model from client #{client_idx}")
+            server.update(local_model_dict, client_model_step[client_idx], client_idx)
+            global_update_time = time.time() - global_update_start
+
+            # Remove the completed request from list
+            recv_reqs.pop(client_idx)
+            if global_step < cfg.num_epochs:
+                # Convert the updated model to bytes
+                global_model = server.model.state_dict()
+                gloabl_model_buffer = io.BytesIO()
+                torch.save(global_model, gloabl_model_buffer)
+                global_model_bytes = gloabl_model_buffer.getvalue()
+
+                # Send (buffer size, finish flag) - INFO - to the client in a blocking way
+                comm.send((len(global_model_bytes), False), dest=client_idx+1, tag=client_idx+1)
+
+                # Send the buffered model - MODEL - to the client in a NON-blocking way
+                comm.Send(np.frombuffer(global_model_bytes, dtype=np.byte), dest=client_idx+1, tag=client_idx+1+comm_size) 
```

Comment:
Is this supposed to be `Isend`?

Reply:
Thanks for this catch! Let me double check this part

Reply:
I double check the code and find there is a typo in the comment. It should be a blocking send.